self_enrollment_tool/utils.py

```diff
@@ -0,0 +1,76 @@
+import logging
+
+from canvas_sdk import RequestContext
+from canvas_sdk.exceptions import CanvasAPIError
+from canvas_sdk.methods.external_tools import (create_external_tool_courses,
+                                               list_external_tools_courses)
+from canvas_sdk.utils import get_all_list_data
+from django.conf import settings
+from icommons_common.canvas_api.helpers import \
+    roles as canvas_api_helpers_roles
+from icommons_common.models import UserRole
+
+logger = logging.getLogger(__name__)
+
+SDK_SETTINGS = settings.CANVAS_SDK_SETTINGS
+# make sure the session_inactivity_expiration_time_secs key isn't in the settings dict
+SDK_SETTINGS.pop('session_inactivity_expiration_time_secs', None)
+SDK_CONTEXT = RequestContext(**SDK_SETTINGS)
+
+
+def get_canvas_roles():
+    logger.debug(f"called _get_canvas_roles()")
+    roles = []
+    try:
+        canvas_roles = canvas_api_helpers_roles.get_roles_for_account_id('self')
+        user_roles = UserRole.objects.filter(pk__in=settings.SELF_ENROLLMENT_TOOL_ROLES_LIST)
+        for role in user_roles:
+            label = canvas_roles[role.canvas_role_id]['label']
+            roles.append({'roleId': role.role_id, 'roleName': label})
+    except CanvasAPIError:
+        logger.exception("Failed to retrieve roles for the root account from Canvas API")
+        return None
+    except Exception:
+        logger.exception("Unhandled exception in _get_canvas_roles; aborting.")
+        return None
+
+    roles.sort(key=lambda x: x['roleName'])
+    logger.debug(f"returning roles: {roles}")
+    return roles
+
+
+def install_unenrollment_tool(course_instance_id, client_id):
+    # make sure that the unenrollment tool is installed
+
+
+    if not client_id:
+        raise Exception('Could not install the self-unenrollment tool into the Canvas course site: the tool client_id is missing.')
+
+    # client_id is in global format (e.g. 18750000000000001)
+    # convert it to local format so we can compare it to the developer_key_id
+    local_client_id = int(client_id) % 10000000000000
+
+    # first, check to see if this tool is already installed
+    installed_tools = get_all_list_data(SDK_CONTEXT, list_external_tools_courses, course_id=f'sis_course_id:{course_instance_id}')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;course_instance_id&#39; (index: 0 | type: Unknown). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;client_id&#39;. The same parameter type is validated here for example: canvas_account_admin_tools/canvas_account_admin_tools/views.py:189. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness, emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) using single or double underscore.

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).